### DISCUSSION TITLE
Download machine.lic

### DISCUSSION THREAD

**Discussion by flopp999** [keygen-sh/community#79] [Q&A] [answered]
I found in the API docs how to be able to download a machine.lic file. But where do I get the Bearer TOKEN?

I tried to create a new API Token under Access but it doesn't work.

**Comment by flopp999** [comment]
OK I managed to download a machine file.
I created an API Token for the product but it says I should be careful to use this.
Is there any other way to be able to download machine file?

```
import shutil
import requests

res = requests.get(
  "https://api.keygen.sh/v1/accounts/xxxxxxxxxxxx/machines/yyyyyyyyyyyyyy/actions/check-out",stream=True,
  headers={
    "Accept": "application/vnd.api+json",
    "Authorization": "Bearer prod-zzzzzzzzzzzzzzzzzzzzzzzzz"
  }
)

with open('machine.lic', 'wb') as out_file:
  shutil.copyfileobj(res.raw, out_file)

```

**Reply by ezekg** [reply]
Use [license authentication](https://keygen.sh/docs/api/authentication/#license-authentication) to download machine files client-side that belong to a specific license.

**Reply by flopp999** [reply]
Thanks it works.
is it possible to get relationships inside the machine.lic through API?
I tried with below curl and got _"{"meta":{"id":"xxxx-7800-yyyy-aaaa-2ffffff"},"errors":[{"title":"Bad request","detail":"unpermitted parameter","source":{"pointer":"/data"}}]}"_
```
curl -X POST https://api.keygen.sh/v1/accounts/xxxxxx/machines/yyyyy/actions/check-out   -H 'Content-Type: application/vnd.api+json'   -H 'Accept: application/vnd.api+json'   -H 'Authorization: License uuuuuuuu'   -d '{
        "data": {
          "type": "machines",
          "attributes": {
            "fingerprint": "kkkkk",
            "platform": "Linux",
            "name": "mitt namn"
          },
          "relationships": {
            "license": {
              "data": {
                "type": "licenses",
                "id": "lllllllll"
              }
            }
          }
        }
      }'

```

**Reply by ezekg** [reply, marked as answer]
You need to use [the `include` query parameter](https://keygen.sh/docs/api/machines/#machines-check-out-query-include) for that.

```bash
curl 'https://api.keygen.sh/v1/accounts/<account>/machines/<machine>/actions/check-out?include=license' \
  -H 'Authorization: License <key>'
```

**Reply by flopp999** [reply]
Thanks

**Reply by qhuanguk** [reply]
I am doing a similar request to download manine.lic file and trying to authorize with license ("-H 'Authorization: License uuuuuuu'"). However, I got an "Access Denied" response as follows:

{"meta":{"id":"942803eb-f552-49b0-bbd1-509d6f12b416"},"errors":[{"title":"Access denied","detail":"You do not have permission to complete the request (license lacks permission to perform action)"}]}

Is this related to some of my settings within the policy (I already tried to set "Authentication Strategy" as "License Keys" and "Mixed" with no success)  and/or license? Thanks

**Reply by ezekg** [reply]
@qhuanguk are you sure the machine you're checking out belongs to the license you're authenticating as?

**Reply by qhuanguk** [reply]
Thanks @ezekg. Honestly I don't know exactly what "belongs to" means and how such belonging is defined within kengen.

I am trialing keygen at the moment for a licensing solution which supports node lock as well as features/entitlements licensing, mainly for offline usage. I am manually creating license on keygen web GUI, and then machine. While creating machine, I select the created license from the drop down list in the relationships part (not sure if this defines the belonging needed). Obviously I could check out the machine file from the GUI, but I am also looking at means to "checkout" this file with the previous curl command over internet. 

I have no problem to check out the file when authorizing with product or admin tokens with a command like
```shell
curl 'https://api.keygen.sh/v1/accounts/{account}/machines/{machine}/actions/check-out?include=license,license.entitlements,license.product&encrypt=1' \  
   -H 'Accept: application/vnd.api+json' \
   -H 'Authorization: Bearer prod-***********'
``` 
However, I don't want to exposure product/admin tokens for the checkout and hence looking for the license key authorization, and that is where I have issues.

Any advise would be greatly appreciated.

**Reply by ezekg** [reply]
When activating a machine, you have to specify a license relationship. This relationship "associates" the machine to that particular license. The machine "belongs to" that license. This means that only that license (or an admin/product) can checkout the machine. Other licenses cannot checkout machines that belong to other licenses. They can only checkout their own.

The fact that you can checkout the machine as a product indicates that you're attempting to checkout Machine A for License X when you're actually authenticating as License Y, who owns Machine B (i.e. they don't own Machine A).

**Reply by ezekg** [reply]
I see another potential problem, and it's likely what's causing the authorization issue. By default, licenses cannot read the product object. So you'd need to remove the `license.product` query parameter. If you're on an Ent tier or using Keygen EE, you could also [customize the license's permissions](https://keygen.sh/docs/api/authorization/#authorization-permissions) to allow the `product.read` action.

**Reply by qhuanguk** [reply]
Indeed the issue lies with the "license.product" query parameter. The checkout works after it is removed. This should work for my case, and will look into customizing the permissions when necessary. Thank you very much, @ezekg

**Reply by ezekg** [reply]
Good to hear. We'll try to make the authorization-related error message clearer.
